src/include/simeng/arch/riscv/ExceptionHandler.hh

```diff
@@ -11,77 +11,60 @@ namespace arch {
 
+  /** Register the instruction which contains the exception with the exception
+   * handler. */
+  void registerException(
+      std::shared_ptr<simeng::Instruction> instruction) override;
+
+  /** Process the result of a syscall from the SyscallHandler. */
+  void processSyscallResult(
+      const simeng::OS::SyscallResult syscallResult) override;
+
   /** Retrieve the results of the handled exception. */
   const ExceptionResult& getResult() const override;
 
  private:
   /** Prints a description of the exception and the instruction that generated
    * it. */
-  void printException(const Instruction& insn) const;
+  void printException() const;
 
   /** The initial handling logic. Returns `true` if no further cycles are
    * required or `false` otherwise, in which case `resumeHandling_` has been set
    * to the next step. */
-  bool init();
-
-  /** Attempt to read a string of max length `maxLength` from address `address`
-   * into the supplied buffer, starting from character `offset`. An offset of
-   * `-1` (default) will queue a read operation for the first character.
-   *
-   * This function will repeatedly set itself as the handler for the next cycle
-   * until it either reads a null character or reaches the maximum length, at
-   * which point it will call `then`, supplying the length of the read string.
-   */
-  bool readStringThen(char* buffer, uint64_t address, int maxLength,
-                      std::function<bool(size_t length)> then, int offset = -1);
-
-  /** Read `length` bytes of data from `ptr`, and then call `then`.
-   *
-   * This function will repeatedly set itself as the handler for the next cycle
-   * until it has read `length` bytes of data. The data may be read in chunks if
-   * it is larger than can be read in a single memory request. The data will be
-   * appended to the member vector `dataBuffer`.
-   */
-  bool readBufferThen(uint64_t ptr, uint64_t length, std::function<bool()> then,
-                      bool firstCall = true);
-
-  /** A data buffer used for reading data from memory. */
-  std::vector<uint8_t> dataBuffer;
-
-  /** Performs a readlinkat syscall using the path supplied. */
-  void readLinkAt(span<char> path);
+  bool initException();
 
   /** Conclude a syscall, setting the return address and state change in the
    * exception results. */
-  bool concludeSyscall(ProcessStateChange& stateChange);
+  bool concludeSyscall();
 
   /** Sets a generic fatal result and returns true. */
   bool fatal();
 
+  /** Resets the state of the exception handler so that it's ready to process
+   * the next instruction. */
+  void resetState();
+
   /** The instruction generating an exception. */
-  const Instruction& instruction_;
+  std::shared_ptr<Instruction> instruction_ = nullptr;
 
   /** The core model object. */
-  const Core& core;
+  const Core& core_;
 
-  /** The process memory. */
-  MemoryInterface& memory_;
+  /** Whether an active syscall result has been received. */
```

Comment:
Same as [this](https://github.com/UoB-HPC/SimEng/pull/285/files#r1114142855) comment

Reply:
Dealt with similarly to referenced comment